mlir/include/mlir/Dialect/Bufferization/IR/SubsetOpInterface.td

```diff
@@ -0,0 +1,159 @@
+    - Extraction flavor: Ops that define a tensor subset. They extract a
+      specified subset from a tensor. There is currently no op interface for
+      such ops. Example: "tensor.extract_slice"
+
+    This interface provides helper methods for efficient bufferization of
+    subset-based tensor IR. Tensor subsets can bufferize to buffer "views"/
+    "aliases" (in contrast to one or multiple less efficient buffer allocation).
+
+    This interface is queried by One-Shot Bufferize to detect cases where a
+    seeming read-after-write is not actually a conflict because the respective
+    ops are operating on equivalent subsets. More details can be found in the
+    documentation of One-Shot Analysis.
+
+    Note: This interface currently assumes that a subset op inserts a single
+    tensor (source) into a destination tensor at a single subset.
+  }];
+  let cppNamespace = "::mlir::bufferization";
+  let methods = [
+      InterfaceMethod<
+        /*desc=*/[{
+          Return the source tensor operand.
+        }],
+        /*retType=*/"::mlir::OpOperand &",
+        /*methodName=*/"getSourceOperand",
+        /*args=*/(ins)
+      >,
+      InterfaceMethod<
+        /*desc=*/[{
+          Return the destination tensor operand.
+        }],
+        /*retType=*/"::mlir::OpOperand &",
+        /*methodName=*/"getDestinationOperand",
+        /*args=*/(ins),
+        /*methodBody=*/"",
+        /*defaultImplementation=*/[{
+          return ::mlir::bufferization::detail::defaultGetDestinationOperand(
+              $_op.getOperation());
+        }]
+      >,
+      InterfaceMethod<
+        /*desc=*/[{
+          Return "true" if this operation inserts into a subset that is
+          equivalent to the subset defined by `candidate`.
+
+          Two subsets are "equivalent" and "same" if they can bufferize to the
+          same buffer views/aliases. If they are "equivalent", the tensor IR
+          may be expressed in terms of different SSA values.
+
+          Example:
+          ```
+          // The subset of the SubsetOpInterface op %1 is equivalent to the
+          // subset defined by %2 (but not "same"):
+          %0 = arith.select %c, %t, %t : tensor<?xf32>
+          %1 = tensor.insert_slice %x into %0[0][5][1]
+              : tensor<5xf32> into tensor<?xf32>
+          %2 = tensor.extract_slice %t[0][5][1] : tensor<?xf32> to tensor<5xf32>
+
+          // The subset of the SubsetOpInterface op %1 is equivalent to and
+          // "same" as the subset defined by %2.
+          %1 = tensor.insert_slice %x into %t[0][5][1]
+              : tensor<5xf32> into tensor<?xf32>
+          %2 = tensor.extract_slice %t[0][5][1] : tensor<?xf32> to tensor<5xf32>
+          ```
+        }],
+        /*retType=*/"bool",
+        /*methodName=*/"isEquivalentSubset",
+        /*args=*/(ins
+            "::mlir::Value":$candidate,
+            "::llvm::function_ref<bool(Value, Value)>":$equivalenceFn)
```

Comment:
can you explain what `equivalenceFn` is in the description.  It's unclear to me whether the main logic for determining equivalent subset should be in `equivalenceFn` or `isEquivalentSubset`.   Maybe I'm just missing something, but seems awkward to have both

Reply:
I expanded the description in #65619. `equivalenceFn` returns "true" if two values are equivalent. When used in One-Shot Bufferize, I use `AnalysisState::areEquivalentBufferizedValues` as `equivalenceFn`. I didn't want to make this interface dependent on other bufferization stuff (such as `AnalysisState`), so that it can be moved to `mlir/Interfaces` in the future.

Reply:
there's context I'm missing since I'm new to some of this stuff, so I won't harp on this much longer.  But I guess I still don't see the need for the hook.  Why can't `isEquivalentSubset` be implemented without it? For example, the default could be
```
bool isEquivalentSubset(Value candidate) {
  return getSubset() == candidate;
```
(sorry if this is a naive question)

Reply:
This would work for `bufferization.materialize_in_destination` because the subset is simply the destination tensor. But `getSubset` may build new IR. E.g., for `tensor.insert_slice`, it will build a new `tensor.extract_slice` op and return that one. I'm going to rename `getSubset` to `buildSubset`.